toolchain/check/subst.cpp

```diff
@@ -389,12 +390,17 @@ class SubstConstantCallbacks final : public SubstInstCallbacks {
   }
 
   // Rebuilds an instruction by building a new constant.
-  auto Rebuild(SemIR::InstId /*old_inst_id*/, SemIR::Inst new_inst) const
+  auto Rebuild(SemIR::InstId old_inst_id, SemIR::Inst new_inst) const
       -> SemIR::InstId override {
-    auto result_id = TryEvalInst(*context_, SemIR::InstId::None, new_inst);
-    CARBON_CHECK(result_id.is_constant(),
+    auto result_id = GetOrAddInst(
```

Comment:
For my understanding, why does this caller use GetOrAddInst() instead of AddInstInNoBlock(), as in impl lookup? Is this why we're getting a bunch of new unused instructions in the semir?

Reply:
Yes. I think we could probably change `GetOrAddInst` to create a new instruction in fewer cases -- perhaps we could distinguish the case where the instruction needs an `InstId` for evaluation purposes only from the case where the instruction needs an `InstId` because it becomes part of the result in some way, and make `GetOrAddInst` do the equivalent of `AddInstInNoBlock` in the former case.